File: app/src/main/java/social/entourage/android/sidemenu/SideMenuItemView.java
package social.entourage.android.sidemenu;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.TypedArray;
import android.support.annotation.StringRes;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.ImageViewCompat;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import social.entourage.android.R;

/**
 * Class that controls a sidemenu item
 */
public class SideMenuItemView extends RelativeLayout {

    private TextView titleView = null;

    public SideMenuItemView(Context context) {
        super(context);
        init(null, 0);
    }

    public SideMenuItemView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs, 0);
    }

    public SideMenuItemView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(attrs, defStyle);
    }

    private void init(AttributeSet attrs, int defStyle) {
        inflate(getContext(), R.layout.layout_side_menu_item, this);
        // Load attributes
        final TypedArray styledAttributes = getContext().obtainStyledAttributes(
                attrs, R.styleable.SideMenuItemView, defStyle, 0);

        try {
            //Handle divider
            boolean bShowDivider = styledAttributes.getBoolean(R.styleable.SideMenuItemView_showDivider, true);
            if (!bShowDivider) {
                ImageView divider = findViewById(R.id.side_menu_item_divider);
                divider.setVisibility(View.GONE);
            }

            //Icon
            ImageView iconView = findViewById(R.id.side_menu_item_icon);
            int iconResourceID = styledAttributes.getResourceId(R.styleable.SideMenuItemView_android_icon, 0);
            if (iconResourceID == 0) {
                iconView.setVisibility(View.GONE);
            } else {
                iconView.setImageResource(iconResourceID);
            }

            int iconTint = styledAttributes.getColor(R.styleable.SideMenuItemView_iconTint, 0);
            if (iconTint != 0) {
                ImageViewCompat.setImageTintList(iconView, ColorStateList.valueOf(iconTint));
            }

            //Title
            titleView = findViewById(R.id.side_menu_item_title);
            String title = styledAttributes.getString(R.styleable.SideMenuItemView_android_title);
            titleView.setText(title);

            int defaultTextColor = titleView.getCurrentTextColor();
            int textColor = styledAttributes.getResourceId(R.styleable.SideMenuItemView_android_textColor, defaultTextColor);
            if (textColor != defaultTextColor) {
                titleView.setTextColor(ContextCompat.getColor(getContext(), textColor));
            }

            float textSize = styledAttributes.getDimensionPixelSize(R.styleable.SideMenuItemView_android_textSize, 0);
            if (textSize > 0) {
                titleView.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize);
            }

            int backgroundResourceID = styledAttributes.getResourceId(R.styleable.SideMenuItemView_textBackground, 0);
            if (backgroundResourceID != 0) {
                titleView.setBackgroundResource(backgroundResourceID);
            }

            boolean centerText = styledAttributes.getBoolean(R.styleable.SideMenuItemView_centerText, false);
            if (centerText) {
                titleView.setGravity(Gravity.CENTER);
            }

            //Right arrow
            boolean bShowRightArrow = styledAttributes.getBoolean(R.styleable.SideMenuItemView_showRightArrow, true);
            if (!bShowRightArrow) {
                ImageView rightArrow = findViewById(R.id.side_menu_item_arrow);
                if (rightArrow != null) rightArrow.setVisibility(GONE);
            }
        }
        finally {
            styledAttributes.recycle();
        }
    }

    public void setTitle(@StringRes int resid) {
        if (titleView != null) {
            titleView.setText(resid);
        }
    }

}

File: app/src/main/java/social/entourage/android/sidemenu/SideMenuItemView.kt
package social.entourage.android.sidemenu

import android.content.Context
import android.content.res.ColorStateList
import android.util.AttributeSet
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.widget.ImageView
import android.widget.RelativeLayout
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import androidx.core.widget.ImageViewCompat
import kotlinx.android.synthetic.entourage.layout_side_menu_item.view.*
import social.entourage.android.R

/**
 * Class that controls a sidemenu item
 */
class SideMenuItemView : RelativeLayout {
    constructor(context: Context?) : super(context) {
        init(null, 0)
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init(attrs, 0)
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle) {
        init(attrs, defStyle)
    }

    private fun init(attrs: AttributeSet?, defStyle: Int) {
        View.inflate(context, R.layout.layout_side_menu_item, this)
        // Load attributes
        val styledAttributes = context.obtainStyledAttributes(
                attrs, R.styleable.SideMenuItemView, defStyle, 0)
        try {
            //Handle divider
            val bShowDivider = styledAttributes.getBoolean(R.styleable.SideMenuItemView_showDivider, true)
            if (!bShowDivider) {
                side_menu_item_divider.visibility = View.GONE
            }

            //Icon
            val iconResourceID = styledAttributes.getResourceId(R.styleable.SideMenuItemView_android_icon, 0)
            if (iconResourceID == 0) {
                side_menu_item_icon.visibility = View.GONE
            } else {
                side_menu_item_icon.setImageResource(iconResourceID)
            }
            val iconTint = styledAttributes.getColor(R.styleable.SideMenuItemView_iconTint, 0)
            if (iconTint != 0) {
                ImageViewCompat.setImageTintList(side_menu_item_icon, ColorStateList.valueOf(iconTint))
            }

            //Title
            val title = styledAttributes.getString(R.styleable.SideMenuItemView_android_title)
            side_menu_item_title.setText(title)
            val defaultTextColor = side_menu_item_title.getCurrentTextColor()
            val textColor = styledAttributes.getResourceId(R.styleable.SideMenuItemView_android_textColor, defaultTextColor)
            if (textColor != defaultTextColor) {
                side_menu_item_title.setTextColor(ContextCompat.getColor(context, textColor))
            }
            val textSize = styledAttributes.getDimensionPixelSize(R.styleable.SideMenuItemView_android_textSize, 0).toFloat()
            if (textSize > 0) {
                side_menu_item_title.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize)
            }
            val backgroundResourceID = styledAttributes.getResourceId(R.styleable.SideMenuItemView_textBackground, 0)
            if (backgroundResourceID != 0) {
                side_menu_item_title.setBackgroundResource(backgroundResourceID)
            }
            val centerText = styledAttributes.getBoolean(R.styleable.SideMenuItemView_centerText, false)
            if (centerText) {
                side_menu_item_title.setGravity(Gravity.CENTER)
            }

            //Right arrow
            val bShowRightArrow = styledAttributes.getBoolean(R.styleable.SideMenuItemView_showRightArrow, true)
            if (!bShowRightArrow) {
                val rightArrow = findViewById<ImageView>(R.id.side_menu_item_arrow)
                if (rightArrow != null) rightArrow.visibility = View.GONE
            }
        } finally {
            styledAttributes.recycle()
        }
    }

    fun setTitle(@StringRes resid: Int) {
        side_menu_item_title?.setText(resid)
    }
}